src/Bitmap_cubical_complex/include/gudhi/Persistence_on_rectangle.h

```diff
@@ -0,0 +1,595 @@
+  // To handle boundaries correctly, we imagine a single exterior cell with filtration +inf.
+  template<class Out>
+  void dual(Out&&out){
+    for (auto e : boost::adaptors::reverse(edges)) {
+      dualize_edge(e);
+      Index a = ds_find_set_square(e.v1);
+      Index b = ds_find_set_square(e.v2);
+      GUDHI_CHECK(a != b, std::logic_error("Bug in Gudhi"));
+      // This is more robust in case the input contains inf? I used to set the filtration of 0 to inf.
+      if (b == 0 || (a != 0 && input(a) < input(b))) std::swap(a, b);
+      ds_parent_square(b) = a;
+      if constexpr (output_index)
+        out(e.f.out(), b);
+      else
+        out(e.f.out(), input(b));
+    }
+  }
+};
+// Ideas for improvement:
+// * for large hard (many intervals) inputs, primal/dual dominate the running time because of the random reads in find_set and input(a/b). The input load would be cheaper if we stored it with parents, but then find_set would be slower.
+// * to increase memory locality, maybe pairing, which is currently arbitrary, could use some heuristic to favor some pairs over others.
+// * try to loosen tight dependency chains, load values several instructions before they are needed. Performing 2 find_set in lock step surprisingly doesn't help.
+
+/**
+ * @private
+ * Compute the persistence diagram of a function on a 2d cubical complex, defined as a lower-star filtration of the values at the top-dimensional cells.
+ *
+ * @tparam output_index If false, each argument of the out functors is a filtration value. If true, it is instead the index of this filtration value in the input.
+ * @tparam Filtration_value Must be comparable with `operator<`.
+ * @tparam Index This is used to index the elements of `input`, so it must be large enough to represent the size of `input`.
+ * @param[in] input Pointer to `nrows*n_cols` filtration values for the square cells. Note that the values are assumed to be stored in C order, unlike `Gudhi::cubical_complex::Bitmap_cubical_complex` (you can exchange `n_rows` and `n_cols` for compatibility).
+ * @param[in] n_rows number of rows of `input`.
+ * @param[in] n_cols number of columns of `input`.
+ * @param[out] out0 For each interval (b, d) in the persistence diagram of dimension 0, the function calls `out0(b, d)`.
+ * @param[out] out1 Same as `out0` for persistence in dimension 1.
+ * @returns The global minimum, which is not paired and is thus the birth of an infinite persistence interval of dimension 0.
+ */
+template <bool output_index = false, typename Filtration_value, typename Index, typename Out0, typename Out1>
+auto persistence_on_rectangle_from_top_cells(Filtration_value const* input, Index n_rows, Index n_cols, Out0&&out0, Out1&&out1){
+#ifdef GUDHI_DETAILED_TIMES
+  Gudhi::Clock clock;
+#endif
+  GUDHI_CHECK(n_rows >= 2 && n_cols >= 2, std::domain_error("The complex must truly be 2d, i.e. at least 2 rows and 2 columns"));
+  Persistence_on_rectangle<Filtration_value, unsigned, output_index> X;
```

Comment:
Why `unsigned` and not `Index`?

Reply:
Hmm, I think I had Index, documented Index, then temporarily used unsigned for a benchmark and forgot to put it back. Fixed.